### PR TITLE
make Opal.set_encoding find encoding aliases once again

### LINKED ISSUE #2181: "unknown encoding name -" with Opal 1.1
Summary: either parser 3.0 or Opal 1.1 changed in a way that likely will prevent https://opalrb.com/try/ from being upgraded to 1.1 (it currently claims to be based on 1.0.3).

Detail:

https://www.ruby2js.com/demo works with Opal 1.0.5.  When upgrading to Opal 1.1, the following is received:

```
ArgumentError: unknown encoding name - ArgumentError: unknown encoding name - 
    at Object.Opal.send2 (http://localhost:4001/demo/ruby2js.js:1729:17)
    at Opal.send (http://localhost:4001/demo/ruby2js.js:1719:17)
    at Function.$exception (http://localhost:4001/demo/ruby2js.js:5208:14)
    at Function.$raise (http://localhost:4001/demo/ruby2js.js:4886:31)
    at Function.$find (http://localhost:4001/demo/ruby2js.js:19529:14)
    at String.$force_encoding (http://localhost:4001/demo/ruby2js.js:19901:43)
    at Function.$reencode_string (http://localhost:4001/demo/ruby2js.js:27953:26)
    at constructor.$source= (http://localhost:4001/demo/ruby2js.js:28037:33)
    at Object.Opal.send2 (http://localhost:4001/demo/ruby2js.js:1729:17)
    at Opal.send (http://localhost:4001/demo/ruby2js.js:1719:17)
```

ruby2js.js:1729 corresponds to the following line:
https://github.com/ruby2js/ruby2js/blob/78c68cb23c196b493c8461ccd12534af567e6210/demo/patch.opal#L14

ruby2js.js:28037 contains:
`input = self.$class().$reencode_string(input);`

If I step into $reencode_string,
  * original_encoding is UTF-8 with Opal 1.0.5
  * original_encoding is nil (original_encoding.constructor is NilClass) with Opal 1.1.0

Note that the code that is failing was literally lifted from Opal:

https://github.com/opal/opal/blob/b2a27d2d263e11db715c1f042a0ce6b9be59105b/lib/opal/parser/patch.rb#L9

Ruby2JS uses the [whitequark parser](https://github.com/whitequark/parser) in substantially the same way as Opal does, with the Opal patches as a starter with a few additions.  Additions, by the way, that Opal is welcome to reuse and/or adapt.

### PR DESCRIPTION
**Pull request by rubys**
Also have set_encoding fail if encoding is not supported

Fixes #2181

### CODE REVIEW

File: spec/filters/unsupported/string.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # NOTE: run bin/format-filters after changing this file
 opal_unsupported_filter "String" do
   fails "BasicObject#__id__ returns a different value for two String literals"
+  fails "Module#const_defined? returns true when passed a constant name with EUC-JP characters"
```
Comment (Mogztter) [inline spec/filters/unsupported/string.rb:4]: Is this related? Do you know why we need to ignore these?

Comment (rubys) [reply spec/filters/unsupported/string.rb:4]: Those tests pass in an encoding that is not supported by Opal.  This patch restores the Opal 1.0.5 behavior of throwing an ArgumentError when an encoding is provided that is not supported.

File: opal/corelib/runtime.js [hunk 2]
```diff
@@ -2445,7 +2445,7 @@
     if (typeof str === 'string')
       throw Opal.FrozenError.$new("can't modify frozen String");
 
-    var encoding = Opal.encodings[name];
+    var encoding = Opal.Encoding.$find(name);
```
Comment (Mogztter) [inline opal/corelib/runtime.js:2448]: Another way to fix this issue is to call `toUpperCase()` on `name`.
For reference, `Opal.encodings` contains:

```js
{
  'UTF-8': constructor {
    '$$id': 14,
    name: 'UTF-8',
    names: [ 'UTF-8', 'CP65001' ],
    ascii: true,
    dummy: false
  },
  CP65001: constructor {
    '$$id': 14,
    name: 'UTF-8',
    names: [ 'UTF-8', 'CP65001' ],
    ascii: true,
    dummy: false
  },
  'UTF-16LE': constructor {
    '$$id': 22,
    name: 'UTF-16LE',
    names: [ 'UTF-16LE' ],
    ascii: false,
    dummy: false
  },
  'UTF-16BE': constructor {
    '$$id': 28,
    name: 'UTF-16BE',
    names: [ 'UTF-16BE' ],
    ascii: false,
    dummy: false
  },
  'UTF-32LE': constructor {
    '$$id': 34,
    name: 'UTF-32LE',
    names: [ 'UTF-32LE' ],
    ascii: false,
    dummy: false
  },
  'ASCII-8BIT': constructor {
    '$$id': 40,
    name: 'ASCII-8BIT',
    names: [ 'ASCII-8BIT', 'BINARY', 'US-ASCII', 'ASCII' ],
    ascii: true,
    dummy: true
  },
  BINARY: constructor {
    '$$id': 40,
    name: 'ASCII-8BIT',
    names: [ 'ASCII-8BIT', 'BINARY', 'US-ASCII', 'ASCII' ],
    ascii: true,
    dummy: true
  },
  'US-ASCII': constructor {
    '$$id': 40,
    name: 'ASCII-8BIT',
    names: [ 'ASCII-8BIT', 'BINARY', 'US-ASCII', 'ASCII' ],
    ascii: true,
    dummy: true
  },
  ASCII: constructor {
    '$$id': 40,
    name: 'ASCII-8BIT',
    names: [ 'ASCII-8BIT', 'BINARY', 'US-ASCII', 'ASCII' ],
    ascii: true,
    dummy: true
  }
}
```

Comment (Mogztter) [reply opal/corelib/runtime.js:2448]: That's a good point :+1:

**Review by elia — APPROVED** [review]
Thanks 👍 
I'll take this and move the implementation of find_encoding to the runtime to jeep things tight, since this is called for every string in a file when the set encoding requires it.

(I'll try to update this PR and then merge it)

**Comment by elia** [comment]
Merged as part of #2190 thanks for catching this @rubys!
